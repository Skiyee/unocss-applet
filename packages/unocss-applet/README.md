<p align="center">
<img src="https://github.com/unocss-applet/unocss-applet/raw/main/public/logo.svg" style="width:100px;" />
<h1 align="center">UnoCSS Applet</h1>
<p align="center">Using <a href="https://github.com/unocss/unocss">UnoCSS</a> in applet(for <a href="https://github.com/dcloudio/uni-app">UniApp</a> and <a href="https://github.com/NervJS/taro">Taro</a>) to be compatible with unsupported syntax.</p>
</p>
<p align="center">
<a href="https://www.npmjs.com/package/unocss-applet"><img src="https://img.shields.io/npm/v/unocss-applet?color=333333&amp;label=" alt="NPM version"></a></p>

## Presets and Plugins

- [unocss-applet](https://github.com/unocss-applet/unocss-applet/tree/main/packages/unocss-applet) - The default package with common presets and plugins
- [@unocss-applet/preset-applet](https://github.com/unocss-applet/unocss-applet/tree/main/packages/preset-applet) - The default preset (right now it's equivalent to `@unocss/preset-uno`)
- [@unocss-applet/preset-rem-to-rpx](https://github.com/unocss-applet/unocss-applet/tree/main/packages/preset-rem-to-rpx) - Coverts rem to rpx for utils.
- [@unocss-applet/transformer-applet](https://github.com/unocss-applet/unocss-applet/tree/main/packages/transformer-applet) - Compile classes that do not support applets into one class.
- [@unocss-applet/transformer-attributify](https://github.com/unocss-applet/unocss-applet/tree/main/packages/transformer-attributify) - Enables Attributify Mode for applet.

## Installation

```bash
npm i unocss-applet --save-dev # with npm
yarn add unocss-applet -D # with yarn
pnpm add unocss-applet -D # with pnpm
```

## Usage

### UnoCSS config

<details>
<summary>unocss.config.ts</summary><br>

```ts
import {
  defineConfig,
  presetAttributify,
  transformerDirectives,
  transformerVariantGroup,
} from 'unocss'

import {
  presetApplet,
  presetRemToRpx,
  transformerApplet,
  transformerAttributify,
} from 'unocss-applet'

export default defineConfig({
  presets: [
    presetApplet(),
    /**
     * you can add `presetAttributify()` here to enable unocss attributify mode prompt
     * although preset is not working for applet, but will generate useless css
     */
    presetAttributify(),
    presetRemToRpx(),
  ],
  transformers: [
    transformerDirectives(),
    transformerVariantGroup(),
    // Don't change the following order
    transformerAttributify(),
    transformerApplet(),
  ],
})
```

<br></details>

### For Platform

<details>
<summary>For UniApp with Vue3 and Vite</summary><br>

`vite.config.ts`

```ts
import UnoCSS from 'unocss/vite'

export default {
  plugins: [
    UnoCSS(),
  ],
}
```

`main.ts`

```ts
import 'uno.css'
```

<br></details>

<details>
<summary>For Taro(v3.5.6) with Vue3 and Webpack5</summary><br>

`config/index.js`

```js
import UnoCSS from 'unocss/webpack'
const config = {
  mini: {
    // ...
    webpackChain(chain, _webpack) {
      chain.plugin('unocss').use(UnoCSS())
    },
  },
  h5: {
    // ...
    webpackChain(chain, _webpack) {
      chain.plugin('unocss').use(UnoCSS())
    },
  },
}
```

`app.ts`

```ts
import 'uno.css'
```

<br></details>

## Example

- [starter-uni](https://github.com/zguolee/starter-uni)
- [ColorTimetable](https://github.com/zguolee/ColorTimetable)

## Acknowledgement

- [UnoCSS](https://github.com/unocss/unocss)

## License

MIT License &copy; 2022-PRESENT [Neil Lee](https://github.com/zguolee)
