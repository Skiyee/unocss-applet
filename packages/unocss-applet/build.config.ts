import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  entries: [
    'src/index',
    'src/preset-applet',
    'src/preset-rem-to-rpx',
    'src/transformer-rename-class',
  ],
  declaration: true,
  clean: true,
  rollup: {
    emitCJS: true,
  },
  externals: [
    'unocss',
  ],
})
