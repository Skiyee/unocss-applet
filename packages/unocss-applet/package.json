{
  "name": "unocss-applet",
  "type": "module",
  "version": "0.0.10",
  "packageManager": "pnpm@7.2.1",
  "description": "",
  "author": "Neil Lee <zguolee@foxmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/unocss-applet/unocss-applet#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss-applet/unocss-applet.git"
  },
  "bugs": "https://github.com/unocss-applet/unocss-applet/issues",
  "keywords": [
    "unocss",
    "uni-app",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs"
    },
    "./preset-applet": {
      "types": "./dist/preset-applet.d.ts",
      "require": "./dist/preset-applet.cjs",
      "import": "./dist/preset-applet.mjs"
    },
    "./preset-rem-to-rpx": {
      "types": "./dist/preset-rem-to-rpx.d.ts",
      "require": "./dist/preset-rem-to-rpx.cjs",
      "import": "./dist/preset-rem-to-rpx.mjs"
    },
    "./transformer-rename-class": {
      "types": "./dist/transformer-rename-class.d.ts",
      "require": "./dist/transformer-rename-class.cjs",
      "import": "./dist/transformer-rename-class.mjs"
    }
  },
  "main": "./dist/index.mjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub"
  },
  "dependencies": {
    "@unocss-applet/preset-applet": "workspace:*",
    "@unocss-applet/preset-rem-to-rpx": "workspace:*",
    "@unocss-applet/transformer-rename-class": "workspace:*"
  }
}
