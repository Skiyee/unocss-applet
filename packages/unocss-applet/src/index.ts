export { default as presetApplet } from '@unocss-applet/preset-applet'
export { default as presetRemRpx } from '@unocss-applet/preset-rem-rpx'
export { default as transformerAttributify } from '@unocss-applet/transformer-attributify'
export { default as transformerApplet } from '@unocss-applet/transformer-applet'
// deprecated from v0.3.0
export { default as presetRemToRpx } from '@unocss-applet/preset-rem-to-rpx'
