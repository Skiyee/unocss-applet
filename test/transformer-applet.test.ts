import { promises as fs } from 'node:fs'
import path from 'node:path'
import { describe, expect, test } from 'vitest'
import type { UnoGenerator } from '@unocss/core'
import { createGenerator } from '@unocss/core'
import MagicString from 'magic-string'
import transformerApplet from '@unocss-applet/transformer-applet'
import { transformerDirectives, transformerVariantGroup } from 'unocss'
import { presetApplet } from 'unocss-applet'

describe('transformer-applet', () => {
  const uno = createGenerator({
    presets: [
      presetApplet(),
    ],
    transformers: [
      transformerDirectives(),
      transformerVariantGroup(),
    ],
  })
  const transformer = transformerApplet()

  async function transform(code: string, _uno: UnoGenerator = uno) {
    const s = new MagicString(code)
    const { css } = await uno.generate(code, { preflights: false })

    await transformer.transform(s, 'foo.js', { uno: _uno, tokens: new Set() } as any)
    const result = s.toString()
    return {
      code: result,
      css,
    }
  }

  test('basic', async () => {
    const content = await fs.readFile(
      path.resolve(__dirname, './fixtures/rules.vue'),
    )
    const result = await transform(content.toString())
    expect(result).toMatchSnapshot()
  })
})
