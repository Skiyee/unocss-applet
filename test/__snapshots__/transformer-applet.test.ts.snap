// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformer-applet > basic 1`] = `
{
  "code": "<!-- eslint-disable @typescript-eslint/quotes -->
<script setup lang=\\"ts\\">
import { ref } from 'vue'
const bg = 'bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_'

const icon = \\"i-carbon:campsite\\"
const bgIgnore = 'bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_'
const index = 1
const customClass = 'text-red'
const bool = ref<boolean>()
const border = ref(true)
const fixed = ref(true)
</script>

<template>
  <div class=\\"text-center aaa\\" p=\\"4\\">
    <div text=\\"4xl\\" class=\\"rotate-180 i-carbon-campsite\\" :class=\\"bg\\" />
    <div :class=\\"[border && 'u-border-top', fixed && 'u-tabbar--fixed']\\" />
    <div class=\\"border bg-blue-200 font-(light mono) \\" :cc=\\"customClass\\">
      <div class=\\"hover:(!bg-gray-600 text-red font-bold)\\" text=\\"#fff\\">
        {{ 'hover:(!bg-gray-600 text-red font-bold)' }}
      </div>
    </div>
    <div :class=\\"\`p-2_a_5 \${bool ? 'p-0_a_5' : ''}\`\\" m-2 :hover-class=\\"['_a_bg-green']\\">
      class=\\"hover_a_bg-green\\"
    </div>
    <div flex=\\"~ col gap-1\\" class=\\"p-1\\" items-center :class=\\"bool ? 'text-yellow-500 px-2_a_5' : ''\\">
      <div :class=\\"icon\\" inline-block color=\\"blue\\" text=\\"xl !red\\" />
      <div bg=\\"green-(!200 800)\\">
        {{ \`index\${index + 1}\` }}{{ \`index\` }}
      </div>
    </div>
    <div flex=\\"~ col\\" un-b=\\"~ green\\" :class=\\"\`bg-\${bg}\`\\">
      <div text-right h-10 flex=\\"1\\" text=\\"red\\" :class=\\"{ 'text-sm': index > 0 }\\">
        0123456789
      </div>
      <div
        h-10 flex=\\"1\\" :class=\\"[index > 1 ? 'text' : '']\\"
        text=\\"blue dark:(red !bold)\\" :style=\\"[index > 1 ? '' : '']\\"
        :type=\\"index > 1\\"
      >
        {{ bgIgnore }}
      </div>
    </div>
    <div class=\\"bg-_a_url_a_https_a__a__a_static_a_runoob_a_com_a_images_a_demo_a_demo2_a_jpg_a__a_\\" w-40 h-20 ma color=\\"white\\" bg=\\"center cover\\">
      {{ 'bg-_a_url_a_https_a__a__a_static_a_runoob_a_com_a_images_a_demo_a_demo2_a_jpg_a__a_' }}
    </div>
    <div class=\\"m-0_a_5 p-1 text-2xl\\" :class=\\"bool ? '' : 'text-yellow-500 p-2_a_5'\\">
      abckefghijklmnopqrstuvwxyz
    </div>
  </div>
</template>
",
  "css": "/* layer: default */
.fixed{position:fixed;}
.m-0_a_5{margin:0.125rem;}
.m-2{margin:0.5rem;}
.ma{margin:auto;}
.inline-block{display:inline-block;}
.h-10{height:2.5rem;}
.h-20{height:5rem;}
.w-40{width:10rem;}
.rotate-180{--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-rotate:180deg;transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
.items-center{align-items:center;}
.gap-1{grid-gap:0.25rem;gap:0.25rem;}
.border{border-width:1px;}
._a_bg-green{--un-bg-opacity:1 !important;background-color:rgba(74,222,128,var(--un-bg-opacity)) !important;}
.bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_{--un-bg-opacity:1;background-color:hsla(2.7,81.9%,69.6%,var(--un-bg-opacity));}
.bg-blue-200{--un-bg-opacity:1;background-color:rgba(191,219,254,var(--un-bg-opacity));}
.hover_a_bg-green:hover{--un-bg-opacity:1;background-color:rgba(74,222,128,var(--un-bg-opacity));}
.bg-_a_url_a_https_a__a__a_static_a_runoob_a_com_a_images_a_demo_a_demo2_a_jpg_a__a_{--un-url:url(https://static.runoob.com/images/demo/demo2.jpg);background-image:var(--un-url);}
.p-0_a_5{padding:0.125rem;}
.p-1{padding:0.25rem;}
.p-2_a_5{padding:0.625rem;}
.px-2_a_5{padding-left:0.625rem;padding-right:0.625rem;}
.text-center{text-align:center;}
.text-right{text-align:right;}
.text-2xl{font-size:1.5rem;line-height:2rem;}
.text-sm{font-size:0.875rem;line-height:1.25rem;}
.text-red{--un-text-opacity:1;color:rgba(248,113,113,var(--un-text-opacity));}
.text-yellow-500{--un-text-opacity:1;color:rgba(234,179,8,var(--un-text-opacity));}",
}
`;
